Libraries/src/Amazon.Lambda.ApplicationLoadBalancerIdentity/ALBIdentityMiddleware.cs

```diff
@@ -0,0 +1,245 @@
+﻿namespace Amazon.Lambda.ApplicationLoadBalancerIdentity
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.IO;
+    using System.Net.Http;
+    using System.Security.Claims;
+    using System.Security.Cryptography;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Extensions.Primitives;
+    using Microsoft.IdentityModel.JsonWebTokens;
+    using Microsoft.IdentityModel.Tokens;
+    using Org.BouncyCastle.Crypto.Parameters;
+    using Org.BouncyCastle.OpenSsl;
+
+    /// <summary>
+    /// A middleware class that extends the ALB OpenId Connect authentication functionality into ASP.NET Core.
+    /// </summary>
+    public class ALBIdentityMiddleware
+    {
+        internal const string OidcDataHeader = "x-amzn-oidc-data";
+        internal const string AWSRegionEnvironmentVariable = "AWS_REGION";
+        internal const string ALBPublicKeyUrlFormatString = "https://public-keys.auth.elb.{0}.amazonaws.com/{1}";
+
+        internal static HttpClient InternalHttpClient;
+
+        private const string NameClaimType = "sub";
+        private const string AuthenticationType = "oidc";
+        private const string ContentTypeHeader = "Content-Type";
+        private const string ContentTypeValue = "text/plain";
+
+        /// Caching related properties
+        private static uint[] byteLookup;
+        private readonly bool cacheIdentities;
+        private readonly MemoryCache cachedIds;
+        private readonly SHA1 sha;
+        private readonly TimeSpan cacheDuration;
+
+        // Validation related properties
+        private readonly string region;
+        private readonly ConcurrentDictionary<string, Task<TokenValidationParameters>> validationParameters;
```

Comment:
If the `Task` fails for an intermittent error the dictionary will always have the failed task in the dictionary. I would prefer only the TokenValidationParameters is in the dictionary and not the Task.

Reply:
Good catch, I will implement your recommendation.